File: app/schemas/aptitude.py
from pydantic import BaseModel, Field, field_validator, model_validator
from typing import Any, Dict, List, Optional, Union
from datetime import datetime
from enum import Enum


class DifficultyEnum(str, Enum):
    EASY = "Easy"
    MEDIUM = "Medium"
    HARD = "Hard"


# ─── Aptitude Attempt ────────────────────────────────────────────────────────

class AptitudeAttemptCreate(BaseModel):
    student_id: str
    topic: str
    subtopic: Optional[str] = None
    score: float = Field(..., ge=0)
    max_score: float = Field(default=100.0, ge=1)
    accuracy: float = Field(..., ge=0, le=100)
    questions_attempted: Optional[int] = Field(default=None, ge=1)
    correct_answers: Optional[int] = Field(default=None, ge=0)
    wrong_answers: Optional[int] = Field(default=None, ge=0)
    skipped_answers: int = Field(default=0, ge=0)
    speed: Optional[float] = Field(default=None, ge=0)
    difficulty: Optional[str] = None
    average_solving_time: Optional[float] = None
    total_time_taken: Optional[float] = None
    difficulty_level: Optional[str] = None
    test_date: Optional[datetime] = None

    @field_validator("accuracy")
    @classmethod
    def validate_accuracy(cls, v: float) -> float:
        return round(v, 2)

    @model_validator(mode="after")
    def normalize_create(self) -> "AptitudeAttemptCreate":
        self.difficulty_level = self.difficulty_level or self.difficulty or DifficultyEnum.MEDIUM.value
        if self.average_solving_time is None and self.speed is not None:
            self.average_solving_time = self.speed
        if self.questions_attempted is None:
            self.questions_attempted = 1
        if self.correct_answers is None:
            self.correct_answers = max(
                0,
                min(self.questions_attempted, round(self.questions_attempted * self.accuracy / 100)),
            )
        if self.wrong_answers is None:
            self.wrong_answers = max(
                0,
                self.questions_attempted - self.correct_answers - self.skipped_answers,
            )
        return self


class AptitudeAttemptResponse(BaseModel):
    id: Union[int, str]
    student_id: str
    topic: str
    subtopic: Optional[str] = None
    score: float
    accuracy: float
    speed: Optional[float] = None
    difficulty: str = "Medium"
    created_at: datetime
    max_score: float = 100.0
    questions_attempted: int = 1
    correct_answers: int = 0
    wrong_answers: int = 0
    average_solving_time: Optional[float] = None
    difficulty_level: str = "Medium"
    test_date: Optional[datetime] = None

    @model_validator(mode="after")
    def normalize_response(self) -> "AptitudeAttemptResponse":
        self.difficulty = self.difficulty or self.difficulty_level or "Medium"
        self.difficulty_level = self.difficulty_level or self.difficulty or "Medium"
        if self.speed is None:
            self.speed = self.average_solving_time
        if self.average_solving_time is None:
            self.average_solving_time = self.speed
        if self.test_date is None:
            self.test_date = self.created_at
        return self


# ─── Topic Progress ───────────────────────────────────────────────────────────

class TopicProgressResponse(BaseModel):
    id: int
    student_id: str
    topic: str
    mastery_score: float
    consistency_score: float
    improvement_trend: float
    total_attempts: int
    average_accuracy: float
    average_speed: float
    readiness_percentage: float
    last_practiced: Optional[datetime] = None
    streak_days: int
    updated_at: datetime

    class Config:
        from_attributes = True


# ─── Analytics ────────────────────────────────────────────────────────────────

class TopicAnalytics(BaseModel):
    topic: str
    mastery_score: float
    average_accuracy: float
    average_speed: float
    total_attempts: int
    improvement_trend: float
    readiness_percentage: float
    is_weak: bool
    is_strong: bool


class OverallAnalyticsResponse(BaseModel):
    student_id: str
    total_attempts: int
    overall_accuracy: float
    overall_speed: float
    overall_readiness: float
    topics: List[TopicAnalytics]
    weak_areas: List[str]
    strong_areas: List[str]
    streak_days: int
    xp_points: int
    badges: List[str]
    last_updated: datetime


# ─── Readiness Score ─────────────────────────────────────────────────────────

class ReadinessScoreResponse(BaseModel):
    student_id: str
    overall_score: float
    quantitative_score: float
    logical_score: float
    verbal_score: float
    data_interpretation_score: float
    puzzles_score: float
    confidence_level: str
    improvement_velocity: float
    company_readiness: Dict[str, float]
    xp_points: int
    badges: List[str]
    current_streak: int
    calculated_at: datetime

    class Config:
        from_attributes = True


# ─── AI Roadmap & Recommendations ────────────────────────────────────────────

class RoadmapRequest(BaseModel):
    student_id: str
    target_companies: Optional[List[str]] = []
    target_readiness: float = Field(default=85.0, ge=50, le=100)
    available_hours_per_day: float = Field(default=2.0, ge=0.5, le=8)


class WeeklyGoal(BaseModel):
    week: int
    topics: List[str] = []
    primary_topic: Optional[str] = None
    target_accuracy: float
    hours_planned: float
    milestones: List[str] = []


class LearningRoadmapResponse(BaseModel):
    student_id: str
    recommended_topics: List[str]
    weekly_goals: List[WeeklyGoal]
    daily_targets: Dict[str, Any]
    focus_areas: List[str]
    completion_progress: float
    ai_insights: Optional[str] = None
    generated_at: datetime

    class Config:
        from_attributes = True


class AIRecommendationResponse(BaseModel):
    student_id: str
    recommendations: List[Dict[str, Any]]
    ai_insights: str
    performance_snapshot: Dict[str, Any]
    priority_topics: List[str]
    daily_goal: str
    generated_at: datetime


# ─── Dashboard ────────────────────────────────────────────────────────────────

class DashboardResponse(BaseModel):
    student_id: str
    readiness_score: float
    xp: int
    streak: int
    total_tests: int
    weak_areas: List[str]
    strong_areas: List[str]
    topic_breakdown: List[TopicAnalytics]
    recent_attempts: List[AptitudeAttemptResponse]
    ai_insight: str
    improvement_trend: List[Dict[str, Any]] = []

File: src/types/aptitude.ts
// ============================================================
// Aptitude Learning Tracker — TypeScript Interfaces
// ============================================================

export type AptitudeTopic =
  | 'Quantitative Aptitude'
  | 'Logical Reasoning'
  | 'Verbal Ability'
  | 'Data Interpretation'
  | 'Puzzles'
  | 'Coding Aptitude';

export type DifficultyLevel = 'Easy' | 'Medium' | 'Hard';
export type ConfidenceLevel = 'Low' | 'Medium' | 'High';
export type RecommendationPriority = 'HIGH' | 'MEDIUM' | 'LOW';

// --- Attempt ---------------------------------------------------

export interface AptitudeAttemptCreate {
  student_id: string;
  topic: AptitudeTopic;
  subtopic?: string;
  total_questions: number;
  correct_answers: number;
  wrong_answers: number;
  skipped_questions?: number;
  total_time_seconds?: number;
  difficulty?: DifficultyLevel;
}

export interface AptitudeAttemptResponse {
  id: string | number;
  student_id: string;
  topic: string;
  subtopic?: string;
  total_questions?: number;
  correct_answers?: number;
  wrong_answers?: number;
  skipped_questions?: number;
  total_time_seconds?: number;
  accuracy?: number;
  score?: number;
  avg_speed?: number;
  speed?: number;
  average_solving_time?: number;
  difficulty?: DifficultyLevel;
  created_at?: string;
}

// --- Topic Progress --------------------------------------------

export interface TopicAnalytics {
  topic: string;
  mastery_score: number;
  average_accuracy: number;
  average_speed: number;
  total_attempts: number;
  improvement_trend: number;
  readiness_percentage: number;
  is_weak: boolean;
  is_strong: boolean;
}

export interface TopicProgressResponse {
  id: number;
  student_id: string;
  topic: string;
  mastery_score: number;
  consistency_score: number;
  improvement_trend: number;
  total_attempts: number;
  average_accuracy: number;
  average_speed: number;
  readiness_percentage: number;
  last_practiced?: string;
  streak_days: number;
  updated_at: string;
}

// --- Overall Analytics -----------------------------------------

export interface OverallAnalyticsResponse {
  student_id: string;
  total_attempts: number;
  overall_accuracy: number;
  overall_speed: number;
  overall_readiness: number;
  topics: TopicAnalytics[];
  weak_areas: string[];
  strong_areas: string[];
  streak_days: number;
  xp_points: number;
  badges: string[];
  last_updated: string;
}

// --- Readiness Score -------------------------------------------

export interface ReadinessScoreResponse {
  student_id: string;
  overall_score: number;
  quantitative_score: number;
  logical_score: number;
  verbal_score: number;
  data_interpretation_score: number;
  puzzles_score: number;
  confidence_level: ConfidenceLevel;
  improvement_velocity: number;
  company_readiness: Record<string, number>;
  xp_points: number;
  badges: string[];
  current_streak: number;
  calculated_at: string;
}

// --- Roadmap ---------------------------------------------------

export interface WeeklyGoal {
  week: number;
  topics: string[];
  primary_topic?: string;
  target_accuracy: number;
  hours_planned: number;
  daily_questions?: number;
  subtopics_to_cover?: string[];
  milestones: string[];
}

export interface LearningRoadmapResponse {
  student_id: string;
  recommended_topics: string[];
  weekly_goals: WeeklyGoal[];
  daily_targets: Record<string, unknown>;
  focus_areas: string[];
  completion_progress: number;
  ai_insights?: string;
  generated_at: string;
}

export interface RoadmapRequest {
  student_id: string;
  target_companies?: string[];
  target_readiness?: number;
  available_hours_per_day?: number;
}

// --- Recommendations -------------------------------------------

export interface Recommendation {
  type: string;
  topic: string;
  action: string;
  priority: RecommendationPriority;
  estimated_gain?: string;
}

export interface RecommendationsResponse {
  student_id: string;
  recommendations: Recommendation[];
  ai_insights: string;
  priority_topics: string[];
  daily_goal: string;
  generated_at: string;
}

// --- Dashboard -------------------------------------------------

export interface ConsistencyEntry {
  date: string;
  count: number;
}

export interface TrendEntry {
  topic: string;
  data: { date: string; accuracy: number }[];
}

export interface TopicSummary {
  topic: string;
  avg_accuracy: number;
  avg_score: number;
  attempts: number;
  is_weak: boolean;
  is_strong: boolean;
}

export interface DashboardResponse {
  student_id: string;
  readiness_score: number;
  xp: number;
  streak: number;
  total_tests: number;
  weak_areas: string[];
  strong_areas: string[];
  topic_breakdown: TopicAnalytics[];
  recent_attempts: AptitudeAttemptResponse[];
  ai_insight: string;
  improvement_trend?: TrendEntry[];
}

// --- Study Plan ------------------------------------------------

export interface StudyPlanDay {
  day: number;
  topic: string;
  target_questions: number;
  is_rest_day: boolean;
  focus: string;
}

export interface StudyPlanResponse {
  student_id: string;
  plan: StudyPlanDay[];
  total_days: number;
}
